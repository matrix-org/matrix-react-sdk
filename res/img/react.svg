<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
