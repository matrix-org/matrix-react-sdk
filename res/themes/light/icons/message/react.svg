<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>