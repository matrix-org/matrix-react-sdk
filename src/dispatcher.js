/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

'use strict';

import flux from "flux";

class MatrixDispatcher extends flux.Dispatcher {
    /**
     * @param {Object|function} payload Required. The payload to dispatch.
     *        If an Object, must contain at least an 'action' key.
     *        If a function, must have the signature (dispatch) => {...}.
     * @param {boolean=} sync Optional. Pass true to dispatch
     *        synchronously. This is useful for anything triggering
     *        an operation that the browser requires user interaction
     *        for.
     */
    dispatch(payload, sync) {
        // Allow for asynchronous dispatching by accepting payloads that have the
        // type `function (dispatch) {...}`
        if (typeof payload === 'function') {
            payload((action) => {
                this.dispatch(action, sync);
            });
            return;
        }

        if (sync) {
            super.dispatch(payload);
        } else {
            // Unless the caller explicitly asked for us to dispatch synchronously,
            // we always set a timeout to do this: The flux dispatcher complains
            // if you dispatch from within a dispatch, so rather than action
            // handlers having to worry about not calling anything that might
            // then dispatch, we just do dispatches asynchronously.
            setTimeout(super.dispatch.bind(this, payload), 0);
        }
    }
}

if (global.mxDispatcher === undefined) {
    global.mxDispatcher = new MatrixDispatcher();
}
export default global.mxDispatcher;
