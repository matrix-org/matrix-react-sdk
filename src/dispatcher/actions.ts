/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

// Dispatcher actions also extend into any arbitrary string, so support that.
export type DispatcherAction = Action | string;

export enum Action {
    // TODO: Populate with actual actions
    // This is lazily generated as it also includes fixing a bunch of references. Work
    // that we don't really want to take on in a giant chunk. We should always define
    // new actions here, and ideally when we touch existing ones we take some time to
    // define them correctly.

    // When defining a new action, please use lower_scored_case with an optional class
    // name prefix. For example, `RoomListStore.view_room` or `view_user_settings`.
    // New definitions should also receive an accompanying interface in the payloads
    // directory.

    /**
     * View a user's profile. Should be used with a ViewUserPayload.
     */
    ViewUser = "view_user",

    /**
     * Open the user settings. No additional payload information required.
     * Optionally can include an OpenToTabPayload.
     */
    ViewUserSettings = "view_user_settings",

    /**
     * Opens the room directory. No additional payload information required.
     */
    ViewRoomDirectory = "view_room_directory",

    /**
     * Forces the theme to reload. No additional payload information required.
     */
    RecheckTheme = "recheck_theme",

    /**
     * Provide status information for an ongoing update check. Should be used with a CheckUpdatesPayload.
     */
    CheckUpdates = "check_updates",

    /**
     * Focuses the user's cursor to the composer. No additional payload information required.
     */
    FocusComposer = "focus_composer",

    /**
     * Opens the user menu (previously known as the top left menu). No additional payload information required.
     */
    ToggleUserMenu = "toggle_user_menu",

    /**
     * Sets the apps root font size. Should be used with UpdateFontSizePayload
     */
    UpdateFontSize = "update_font_size",

    /**
     * Sets a system font. Should be used with UpdateSystemFontPayload
     */
    UpdateSystemFont = "update_system_font",

    /**
     * Changes room based on room list order and payload parameters. Should be used with ViewRoomDeltaPayload.
     */
    ViewRoomDelta = "view_room_delta",

    /**
     * Sets the phase for the right panel. Should be used with SetRightPanelPhasePayload.
     */
    SetRightPanelPhase = "set_right_panel_phase",

    /**
     * Toggles the right panel. Should be used with ToggleRightPanelPayload.
     */
    ToggleRightPanel = "toggle_right_panel",

    /**
     * Trigged after the phase of the right panel is set. Should be used with AfterRightPanelPhaseChangePayload.
     */
    AfterRightPanelPhaseChange = "after_right_panel_phase_change",

    /**
     * Opens the modal dial pad
     */
    OpenDialPad = "open_dial_pad",

    /**
     * Dial the phone number in the payload
     * payload: DialNumberPayload
     */
    DialNumber = "dial_number",

    /**
     * Fired when CallHandler has checked for PSTN protocol support
     * payload: none
     * XXX: Is an action the right thing for this?
     */
    PstnSupportUpdated = "pstn_support_updated",

    /**
     * Similar to PstnSupportUpdated, fired when CallHandler has checked for virtual room support
     * payload: none
     * XXX: Ditto
     */
    VirtualRoomSupportUpdated = "virtual_room_support_updated",

    /**
     * Fired when an upload has started. Should be used with UploadStartedPayload.
     */
    UploadStarted = "upload_started",

    /**
     * Fired when an upload makes progress. Should be used with UploadProgressPayload.
     */
    UploadProgress = "upload_progress",

    /**
     * Fired when an upload is completed. Should be used with UploadFinishedPayload.
     */
    UploadFinished = "upload_finished",

    /**
     * Fired when an upload fails. Should be used with UploadErrorPayload.
     */
    UploadFailed = "upload_failed",

    /**
     * Fired when an upload is cancelled by the user. Should be used with UploadCanceledPayload.
     */
    UploadCanceled = "upload_canceled",

    /**
     * Fired when requesting to join a room
     */
    JoinRoom = "join_room",

    /**
     * Fired when successfully joining a room
     */
    JoinRoomReady = "join_room_ready",

    /**
     * Fired when joining a room failed
     */
    JoinRoomError = "join_room_error",
}
