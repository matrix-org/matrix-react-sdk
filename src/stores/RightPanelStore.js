/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import dis from '../dispatcher/dispatcher';
import {pendingVerificationRequestForUser} from '../verification';
import {Store} from 'flux/utils';
import SettingsStore, {SettingLevel} from "../settings/SettingsStore";
import {RIGHT_PANEL_PHASES, RIGHT_PANEL_PHASES_NO_ARGS} from "./RightPanelStorePhases";

const INITIAL_STATE = {
    // Whether or not to show the right panel at all. We split out rooms and groups
    // because they're different flows for the user to follow.
    showRoomPanel: SettingsStore.getValue("showRightPanelInRoom"),
    showGroupPanel: SettingsStore.getValue("showRightPanelInGroup"),

    // The last phase (screen) the right panel was showing
    lastRoomPhase: SettingsStore.getValue("lastRightPanelPhaseForRoom"),
    lastGroupPhase: SettingsStore.getValue("lastRightPanelPhaseForGroup"),

    // Extra information about the last phase
    lastRoomPhaseParams: {},
};

const GROUP_PHASES = Object.keys(RIGHT_PANEL_PHASES).filter(k => k.startsWith("Group"));

const MEMBER_INFO_PHASES = [
    RIGHT_PANEL_PHASES.RoomMemberInfo,
    RIGHT_PANEL_PHASES.Room3pidMemberInfo,
    RIGHT_PANEL_PHASES.EncryptionPanel,
];

/**
 * A class for tracking the state of the right panel between layouts and
 * sessions.
 */
export default class RightPanelStore extends Store {
    static _instance;

    constructor() {
        super(dis);

        // Initialise state
        this._state = INITIAL_STATE;
    }

    get isOpenForRoom(): boolean {
        return this._state.showRoomPanel;
    }

    get isOpenForGroup(): boolean {
        return this._state.showGroupPanel;
    }

    get roomPanelPhase(): string {
        return this._state.lastRoomPhase;
    }

    get groupPanelPhase(): string {
        return this._state.lastGroupPhase;
    }

    get visibleRoomPanelPhase(): string {
        return this.isOpenForRoom ? this.roomPanelPhase : null;
    }

    get visibleGroupPanelPhase(): string {
        return this.isOpenForGroup ? this.groupPanelPhase : null;
    }

    get roomPanelPhaseParams(): any {
        return this._state.lastRoomPhaseParams || {};
    }

    _setState(newState) {
        this._state = Object.assign(this._state, newState);

        SettingsStore.setValue(
            "showRightPanelInRoom",
            null,
            SettingLevel.DEVICE,
            this._state.showRoomPanel,
        );
        SettingsStore.setValue(
            "showRightPanelInGroup",
            null,
            SettingLevel.DEVICE,
            this._state.showGroupPanel,
        );

        if (RIGHT_PANEL_PHASES_NO_ARGS.includes(this._state.lastRoomPhase)) {
            SettingsStore.setValue(
                "lastRightPanelPhaseForRoom",
                null,
                SettingLevel.DEVICE,
                this._state.lastRoomPhase,
            );
        }
        if (RIGHT_PANEL_PHASES_NO_ARGS.includes(this._state.lastGroupPhase)) {
            SettingsStore.setValue(
                "lastRightPanelPhaseForGroup",
                null,
                SettingLevel.DEVICE,
                this._state.lastGroupPhase,
            );
        }

        this.__emitChange();
    }

    __onDispatch(payload) {
        switch (payload.action) {
            case 'view_room':
            case 'view_group':
                // Reset to the member list if we're viewing member info
                if (MEMBER_INFO_PHASES.includes(this._state.lastRoomPhase)) {
                    this._setState({lastRoomPhase: RIGHT_PANEL_PHASES.RoomMemberList, lastRoomPhaseParams: {}});
                }

                // Do the same for groups
                if (this._state.lastGroupPhase === RIGHT_PANEL_PHASES.GroupMemberInfo) {
                    this._setState({lastGroupPhase: RIGHT_PANEL_PHASES.GroupMemberList});
                }
                break;

            case 'set_right_panel_phase': {
                let targetPhase = payload.phase;
                let refireParams = payload.refireParams;
                // redirect to EncryptionPanel if there is an ongoing verification request
                if (targetPhase === RIGHT_PANEL_PHASES.RoomMemberInfo && payload.refireParams) {
                    const {member} = payload.refireParams;
                    const pendingRequest = pendingVerificationRequestForUser(member);
                    if (pendingRequest) {
                        targetPhase = RIGHT_PANEL_PHASES.EncryptionPanel;
                        refireParams = {
                            verificationRequest: pendingRequest,
                            member,
                        };
                    }
                }
                if (!RIGHT_PANEL_PHASES[targetPhase]) {
                    console.warn(`Tried to switch right panel to unknown phase: ${targetPhase}`);
                    return;
                }

                if (GROUP_PHASES.includes(targetPhase)) {
                    if (targetPhase === this._state.lastGroupPhase) {
                        this._setState({
                            showGroupPanel: !this._state.showGroupPanel,
                        });
                    } else {
                        this._setState({
                            lastGroupPhase: targetPhase,
                            showGroupPanel: true,
                        });
                    }
                } else {
                    if (targetPhase === this._state.lastRoomPhase && !refireParams) {
                        this._setState({
                            showRoomPanel: !this._state.showRoomPanel,
                        });
                    } else {
                        this._setState({
                            lastRoomPhase: targetPhase,
                            showRoomPanel: true,
                            lastRoomPhaseParams: refireParams || {},
                        });
                    }
                }

                // Let things like the member info panel actually open to the right member.
                dis.dispatch({
                    action: 'after_right_panel_phase_change',
                    phase: targetPhase,
                    ...(refireParams || {}),
                });
                break;
            }

            case 'toggle_right_panel':
                if (payload.type === "room") {
                    this._setState({ showRoomPanel: !this._state.showRoomPanel });
                } else { // group
                    this._setState({ showGroupPanel: !this._state.showGroupPanel });
                }
                break;
        }
    }

    static getSharedInstance(): RightPanelStore {
        if (!RightPanelStore._instance) {
            RightPanelStore._instance = new RightPanelStore();
        }
        return RightPanelStore._instance;
    }
}
