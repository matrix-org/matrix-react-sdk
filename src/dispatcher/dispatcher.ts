/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 New Vector Ltd
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { Dispatcher } from "flux";
import { Action } from "./actions";
import { ActionPayload, AsyncActionPayload } from "./payloads";

/**
 * A dispatcher for ActionPayloads (the default within the SDK).
 */
export class MatrixDispatcher extends Dispatcher<ActionPayload> {
    /**
     * Dispatches an event on the dispatcher's event bus.
     * @param {ActionPayload} payload Required. The payload to dispatch.
     * @param {boolean=false} sync Optional. Pass true to dispatch
     *        synchronously. This is useful for anything triggering
     *        an operation that the browser requires user interaction
     *        for. Default false (async).
     */
    dispatch<T extends ActionPayload>(payload: T, sync = false) {
        if (payload instanceof AsyncActionPayload) {
            payload.fn((action: ActionPayload) => {
                this.dispatch(action, sync);
            });
            return;
        }

        if (sync) {
            super.dispatch(payload);
        } else {
            // Unless the caller explicitly asked for us to dispatch synchronously,
            // we always set a timeout to do this: The flux dispatcher complains
            // if you dispatch from within a dispatch, so rather than action
            // handlers having to worry about not calling anything that might
            // then dispatch, we just do dispatches asynchronously.
            setTimeout(super.dispatch.bind(this, payload), 0);
        }
    }

    /**
     * Shorthand for dispatch({action: Action.WHATEVER}, sync). No additional
     * properties can be included with this version.
     * @param {Action} action The action to dispatch.
     * @param {boolean=false} sync Whether the dispatch should be sync or not.
     * @see dispatch(action: ActionPayload, sync: boolean)
     */
    fire(action: Action, sync = false) {
        this.dispatch({action}, sync);
    }
}

export const defaultDispatcher = new MatrixDispatcher();

const anyGlobal = <any>global;
if (!anyGlobal.mxDispatcher) {
    anyGlobal.mxDispatcher = defaultDispatcher;
}

export default defaultDispatcher;
