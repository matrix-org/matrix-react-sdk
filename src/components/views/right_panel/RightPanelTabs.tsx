/*
Copyright 2024 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { useRef } from "react";
import { NavBar, NavItem } from "@vector-im/compound-web";
import { Room } from "matrix-js-sdk/src/matrix";

import { _t } from "../../../languageHandler";
import { RightPanelPhases } from "../../../stores/right-panel/RightPanelStorePhases";
import RightPanelStore from "../../../stores/right-panel/RightPanelStore";
import PosthogTrackers from "../../../PosthogTrackers";
import { useDispatcher } from "../../../hooks/useDispatcher";
import dispatcher from "../../../dispatcher/dispatcher";
import { Action } from "../../../dispatcher/actions";
import SettingsStore from "../../../settings/SettingsStore";
import { UIComponent, UIFeature } from "../../../settings/UIFeature";
import { shouldShowComponent } from "../../../customisations/helpers/UIComponents";
import { useIsVideoRoom } from "../../../utils/video-rooms";

function shouldShowTabsForPhase(phase?: RightPanelPhases): boolean {
    const tabs = [
        RightPanelPhases.RoomSummary,
        RightPanelPhases.RoomMemberList,
        RightPanelPhases.ThreadPanel,
        RightPanelPhases.Extensions,
    ];
    return !!phase && tabs.includes(phase);
}

type Props = {
    room?: Room;
    phase: RightPanelPhases;
};

export const RightPanelTabs: React.FC<Props> = ({ phase, room }): JSX.Element | null => {
    const threadsTabRef = useRef<HTMLButtonElement | null>(null);

    useDispatcher(dispatcher, (payload) => {
        // This actually focuses the threads tab, as its the only interactive element,
        // but at least it puts the user in the right area of the app.
        if (payload.action === Action.FocusThreadsPanel) {
            threadsTabRef.current?.focus();
        }
    });

    const isVideoRoom = useIsVideoRoom(room);

    if (!shouldShowTabsForPhase(phase)) return null;

    return (
        <NavBar className="mx_RightPanelTabs" aria-label="right panel" role="tablist">
            <NavItem
                aria-controls="room-summary-panel"
                id="room-summary-panel-tab"
                onClick={() => {
                    RightPanelStore.instance.pushCard({ phase: RightPanelPhases.RoomSummary }, true);
                }}
                active={phase === RightPanelPhases.RoomSummary}
            >
                {_t("right_panel|info")}
            </NavItem>
            <NavItem
                aria-controls="memberlist-panel"
                id="memberlist-panel-tab"
                onClick={(ev: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
                    RightPanelStore.instance.pushCard({ phase: RightPanelPhases.RoomMemberList }, true);
                    PosthogTrackers.trackInteraction("WebRightPanelRoomInfoPeopleButton", ev);
                }}
                active={phase === RightPanelPhases.RoomMemberList}
            >
                {_t("common|people")}
            </NavItem>
            <NavItem
                aria-controls="thread-panel"
                id="thread-panel-tab"
                onClick={() => {
                    RightPanelStore.instance.pushCard({ phase: RightPanelPhases.ThreadPanel }, true);
                }}
                active={phase === RightPanelPhases.ThreadPanel}
                ref={threadsTabRef}
            >
                {_t("common|threads")}
            </NavItem>
            {SettingsStore.getValue(UIFeature.Widgets) &&
                !isVideoRoom &&
                shouldShowComponent(UIComponent.AddIntegrations) && (
                    <NavItem
                        aria-controls="thread-panel"
                        id="extensions-panel-tab"
                        onClick={() => {
                            RightPanelStore.instance.pushCard({ phase: RightPanelPhases.Extensions }, true);
                        }}
                        active={phase === RightPanelPhases.Extensions}
                    >
                        {_t("common|extensions")}
                    </NavItem>
                )}
        </NavBar>
    );
};
