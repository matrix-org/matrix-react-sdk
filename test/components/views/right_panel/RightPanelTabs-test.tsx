/*
Copyright 2024 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render, fireEvent } from "@testing-library/react";

import dis from "../../../../src/dispatcher/dispatcher";
import RightPanelStore from "../../../../src/stores/right-panel/RightPanelStore";
import { RightPanelPhases } from "../../../../src/stores/right-panel/RightPanelStorePhases";
import { RightPanelTabs } from "../../../../src/components/views/right_panel/RightPanelTabs";
import { Action } from "../../../../src/dispatcher/actions";

describe("<RightPanelTabs />", () => {
    it("Component renders the correct tabs", () => {
        const { container, getByRole } = render(<RightPanelTabs phase={RightPanelPhases.RoomSummary} />);
        expect(container).toMatchSnapshot();

        // We expect Info, People and Threads as tabs
        expect(getByRole("tab", { name: "Info" })).toBeDefined();
        expect(getByRole("tab", { name: "People" })).toBeDefined();
        expect(getByRole("tab", { name: "Threads" })).toBeDefined();
    });

    it("Correct tab is active", () => {
        const { container } = render(<RightPanelTabs phase={RightPanelPhases.RoomMemberList} />);
        expect(container).toMatchSnapshot();
        // Assert that the active tab is Info
        expect(container.querySelectorAll("[aria-selected='true']").length).toEqual(1);
        expect(container.querySelector("[aria-selected='true']")).toHaveAccessibleName("People");
    });

    it("Renders nothing for some phases, eg: FilePanel", () => {
        const { container } = render(<RightPanelTabs phase={RightPanelPhases.FilePanel} />);
        expect(container).toBeEmptyDOMElement();
    });

    it("onClick behaviors work as expected", () => {
        const spy = jest.spyOn(RightPanelStore.instance, "pushCard");
        const { getByRole } = render(<RightPanelTabs phase={RightPanelPhases.RoomSummary} />);

        // Info -> People
        fireEvent.click(getByRole("tab", { name: "People" }));
        expect(spy).toHaveBeenLastCalledWith({ phase: RightPanelPhases.RoomMemberList }, true);

        // People -> Threads
        fireEvent.click(getByRole("tab", { name: "Threads" }));
        expect(spy).toHaveBeenLastCalledWith({ phase: RightPanelPhases.ThreadPanel }, true);

        // Threads -> Info
        fireEvent.click(getByRole("tab", { name: "Info" }));
        expect(spy).toHaveBeenLastCalledWith({ phase: RightPanelPhases.RoomSummary }, true);
    });

    it("Threads tab is focused on action", () => {
        const { getByRole } = render(<RightPanelTabs phase={RightPanelPhases.ThreadPanel} />);
        dis.dispatch({ action: Action.FocusThreadsPanel }, true);
        expect(getByRole("tab", { name: "Threads" })).toHaveFocus();
    });
});
