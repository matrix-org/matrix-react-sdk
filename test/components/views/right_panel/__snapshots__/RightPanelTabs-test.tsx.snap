// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RightPanelTabs /> Component renders the correct tabs 1`] = `
<div>
  <nav
    class="mx_RightPanelTabs _nav-bar_135dy_16"
    role="presentation"
  >
    <ul
      aria-label="right panel"
      class="_nav-bar-items_135dy_22"
      role="tablist"
    >
      <li
        class="_nav-tab_135dy_33"
        data-current="true"
        role="presentation"
      >
        <button
          aria-controls="room-summary-panel"
          aria-selected="true"
          class="_nav-item_135dy_55"
          id="room-summary-panel-tab"
          role="tab"
        >
          Info
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="memberlist-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="memberlist-panel-tab"
          role="tab"
        >
          People
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="thread-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="thread-panel-tab"
          role="tab"
        >
          Threads
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="thread-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="extensions-panel-tab"
          role="tab"
        >
          Extensions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`<RightPanelTabs /> Correct tab is active 1`] = `
<div>
  <nav
    class="mx_RightPanelTabs _nav-bar_135dy_16"
    role="presentation"
  >
    <ul
      aria-label="right panel"
      class="_nav-bar-items_135dy_22"
      role="tablist"
    >
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="room-summary-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="room-summary-panel-tab"
          role="tab"
        >
          Info
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        data-current="true"
        role="presentation"
      >
        <button
          aria-controls="memberlist-panel"
          aria-selected="true"
          class="_nav-item_135dy_55"
          id="memberlist-panel-tab"
          role="tab"
        >
          People
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="thread-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="thread-panel-tab"
          role="tab"
        >
          Threads
        </button>
      </li>
      <li
        class="_nav-tab_135dy_33"
        role="presentation"
      >
        <button
          aria-controls="thread-panel"
          aria-selected="false"
          class="_nav-item_135dy_55"
          id="extensions-panel-tab"
          role="tab"
        >
          Extensions
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
